File: pytests/hc.py
import RPi.GPIO as GPIO
import time

GPIO.setmode(GPIO.BCM)  # Use BCM pin numbering
GPIO.setwarnings(False)

TRIG = 23  # GPIO pin for Trigger
ECHO = 24  # GPIO pin for Echo

print("Distance Measurement In Progress")

# Set up the GPIO pins
GPIO.setup(TRIG, GPIO.OUT)
GPIO.setup(ECHO, GPIO.IN)
print("Waiting For Sensor To Settle")

try:
    while True:
        GPIO.output(TRIG, False)
        time.sleep(2)

        # Send a pulse to the trigger pin
        GPIO.output(TRIG, True)
        time.sleep(0.00001)  # 10µs pulse
        GPIO.output(TRIG, False)

        # Measure the time for the echo response
        while GPIO.input(ECHO) == 0:
            pulse_start = time.time()

        while GPIO.input(ECHO) == 1:
            pulse_end = time.time()

        pulse_duration = pulse_end - pulse_start  # Calculate the pulse duration
        distance = pulse_duration * 17150  # Calculate distance (cm)
        distance = round(distance, 2)  # Round to 2 decimal places

        print("Distance:", distance, "cm")

        time.sleep(0.1)  # Wait 1 second before measuring again

except KeyboardInterrupt:
    print("Stopping distance measurement")
    GPIO.cleanup()  # Clean up GPIO settings

File: cTesting/HC.c
#include <wiringPi.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

#define TRIG 4 // WiringPi pin 4 (GPIO 23)
#define ECHO 5 // WiringPi pin 5 (GPIO 24)

void setup()
{
    wiringPiSetup();       // Initialize WiringPi library
    pinMode(TRIG, OUTPUT); // Set TRIG as output
    pinMode(ECHO, INPUT);  // Set ECHO as input

    // Ensure TRIG is low at startup
    digitalWrite(TRIG, LOW);
    delay(30); // Wait for sensor to settle
}

double measureDistance()
{
    // Send a 10-microsecond pulse to TRIG
    digitalWrite(TRIG, HIGH);
    delayMicroseconds(10);
    digitalWrite(TRIG, LOW);

    // Wait for ECHO to go HIGH
    while (digitalRead(ECHO) == LOW)
        ;

    // Record the start time
    long startTime = micros();

    // Wait for ECHO to go LOW
    while (digitalRead(ECHO) == HIGH)
        ;

    // Record the end time
    long endTime = micros();

    // Calculate the distance in cm
    double duration = endTime - startTime;
    double distance = (duration / 2.0) * 0.0343; // Speed of sound: 343 m/s

    return distance;
}

int main()
{
    setup();

    printf("Ultrasonic Distance Sensor Initialized\n");

    while (1)
    {
        double distance = measureDistance();
        printf("Distance: %.2f cm\n", distance);

        delay(1000); // Wait 1 second before next measurement
    }

    return 0;
}
